consensus/propagation/types.go

```diff
@@ -0,0 +1,230 @@
+package types
+
+import (
+	"errors"
+
+	"github.com/tendermint/tendermint/crypto/merkle"
+	"github.com/tendermint/tendermint/libs/bits"
+	protoprop "github.com/tendermint/tendermint/proto/tendermint/propagation"
+	"github.com/tendermint/tendermint/types"
+)
+
+// TxmetaData keeps track of the hash of a transaction and its location with the
+// protobuf encoded block.
+type TxMetaData struct {
+	Hash  []byte `protobuf:"bytes,1,opt,name=hash,proto3" json:"hash,omitempty"`
+	Start uint32
+	End   uint32
+}
+
+// ToProto converts TxMetaData to its protobuf representation.
+func (t *TxMetaData) ToProto() *protoprop.TxMetaData {
+	return &protoprop.TxMetaData{
+		Hash:  t.Hash,
+		Start: t.Start,
+		End:   t.End,
+	}
+}
+
+// TxMetaDataFromProto converts a protobuf TxMetaData to its Go representation.
+func TxMetaDataFromProto(t *protoprop.TxMetaData) *TxMetaData {
+	return &TxMetaData{
+		Hash:  t.Hash,
+		Start: t.Start,
+		End:   t.End,
+	}
+}
+
+// ValidateBasic checks if the TxMetaData is valid. It fails if Start > End or
+// if the hash is invalid.
+func (t *TxMetaData) ValidateBasic() error {
+	if t.Start > t.End {
+		return errors.New("TxMetaData: Start > End")
+	}
+
+	return types.ValidateHash(t.Hash)
+}
+
+// CompactBlock contains commitments and metadata for reusing transactions that
+// have already been distributed.
+type CompactBlock struct {
+	Height    int64         `json:"height,omitempty"`
+	Round     int32         `json:"round,omitempty"`
+	BpHash    []byte        `json:"bp_hash,omitempty"`
```

Comment:
[nit]
what does bpHash stand for? maybe a comment specifying it's the block propagation hash